### PR TITLE
feat: implement random sort and dynamic timeline sorting via dropdown

### PR DESCRIPTION
**Pull request by akhileshthite**
https://github.com/hyphacoop/reader.distributed.press/pull/13/commits/22310eecfde6c813d57694cc8b12af242745e757

Screenshot 2024-05-09 at 9 15 42 PM

### CODE REVIEW

File: db.js [hunk 1]
```diff
@@ -250,6 +250,31 @@ export class ActivityPubDB extends EventTarget {
...
   }
 
+  async * searchNotesRandom ({ limit = DEFAULT_LIMIT } = {}) {
+    const tx = this.db.transaction(NOTES_STORE, 'readonly')
+    const store = tx.objectStore(NOTES_STORE)
+    const totalNotes = await store.count()
+
+    for (let i = 0; i < limit; i++) {
+      const randomSkip = Math.floor(Math.random() * totalNotes)
+      let cursor = await store.openCursor()
```
Comment (RangerMauve) [inline db.js]: The initial cursor should be opened outside of the for loop

File: db.js [hunk 2]
```diff
@@ -250,6 +250,31 @@ export class ActivityPubDB extends EventTarget {
...
+      if (cursor) {
+        if (randomSkip > 0) { // Only advance if randomSkip is greater than 0
+          await cursor.advance(randomSkip)
+          if (cursor) {
```
Comment (RangerMauve) [inline db.js]: If we reach the end we should open a new cursor.

File: index.html [hunk 3]
```diff
@@ -12,7 +12,17 @@
 </style>
 <div class="container">
   <sidebar-nav></sidebar-nav>
-  <reader-timeline></reader-timeline>
+  <div class="reader-container">
+    <div class="sort-container">
```
Comment (RangerMauve) [inline index.html:16]: Turn this into a custom element so we can keep it cleaner. Have it emit an event for when it changes and add a script on this page to set the `sort` attribute on the timeline element. The timeline element should listen on changes to its sort attribute.

File: timeline.js [hunk 4]
```diff
@@ -56,20 +87,52 @@ class ReaderTimeline extends HTMLElement {
...
-    for await (const note of db.searchNotes({}, { skip: this.skip, limit: this.limit })) {
-      count++
-      this.appendNoteElement(note)
+
+    if (this.sort === 'random' && this.randomNotes.length === 0) {
+      const allNotes = []
+      for await (const note of db.searchNotesRandom(this.limit)) {
+        allNotes.push(note)
+      }
+      this.randomNotes = allNotes.sort(() => Math.random() - 0.5)
```
Comment (RangerMauve) [inline timeline.js]: We shouldn't do any sort if we're getting random notes already

**Review by RangerMauve — CHANGES_REQUESTED** [review]
Nice, good work on the initial pass.

I think we could have a cleaner `hasMore` logic by having the reader track the current count vs how many items total are in the notes collection.

Also instead of storing notes in an array we should render them out as we get them. For duplicates, we can just not bother tracking duplicate notes when rendering the randomized timeline.

Also, we should start doing a page render before ingesting followed accounts to improve initial load.

File: db.js [hunk 5]
```diff
@@ -255,20 +262,23 @@ export class ActivityPubDB extends EventTarget {
     const store = tx.objectStore(NOTES_STORE)
     const totalNotes = await store.count()
```
Comment (RangerMauve) [inline db.js]: I think we can get rid of the uniqueIndexes set and just allow duplicates for now.

File: db.js [hunk 6]
```diff
@@ -255,20 +262,23 @@ export class ActivityPubDB extends EventTarget {
...
+          // Move the cursor to the randomSkip position if not already there
           await cursor.advance(randomSkip)
-          if (cursor) {
-            yield cursor.value
-            cursor = await cursor.continue()
-          }
-        } else {
-          // If randomSkip is 0, yield the first item directly
+        }
+        if (cursor) {
```
Comment (RangerMauve) [inline db.js:231]: We should open a new cursor regardless of wether the previous one was at the end

```suggestion
```

File: timeline.js [hunk 7]
```diff
@@ -124,10 +124,10 @@ class ReaderTimeline extends HTMLElement {
 
   updateIndexes (count) {
```
Comment (RangerMauve) [inline timeline.js]: Mind renaming this to something like "updateHasMore"?

**Review by RangerMauve — CHANGES_REQUESTED** [review]

File: db.js (hunk 5 above)
Comment (akhileshthite) [reply db.js]: But that will be a bad UX :/

File: db.js [hunk 8]
```diff
@@ -220,33 +227,54 @@ export class ActivityPubDB extends EventTarget {
     await tx.done()
   }
 
-  async * searchNotes ({ attributedTo } = {}, { skip = 0, limit = DEFAULT_LIMIT, sort = -1 } = {}) {
+  async * searchNotesRandom ({ limit = DEFAULT_LIMIT } = {}) {
```
Comment (RangerMauve) [inline db.js]: I think you accidentally deleted the searchNotes function. Mind re-adding it and just updating searchNotesRandom.

Alternately maybe we can replace the `sort` argument with `0` or something to indicate `random` and merge the functions?

File: db.js (hunk 5 above)
Comment (RangerMauve) [reply db.js]: Yeah, lets add a todo for later and see how it goes? I think performance and maintenance should come first. I worry our algorithm can get stuck in a loop trying to fill slots at random without duplicates as the amount of data grows. I think "cpu heats up as the timeline is stuck loading" is a likely outcome with how the code was.

**Review by RangerMauve — CHANGES_REQUESTED** [review]

File: timeline.js [hunk 9]
```diff
@@ -95,7 +95,7 @@ class ReaderTimeline extends HTMLElement {
     let count = 0
 
     if (this.sort === 'random') {
-      for await (const note of db.searchNotesRandom({ limit: this.limit })) {
+      for await (const note of db.searchNotes({ limit: this.limit })) {
```
Comment (RangerMauve) [inline timeline.js]: Should this be getting the sort parameter passed in too?

**Review by RangerMauve — CHANGES_REQUESTED** [review]

**Review by RangerMauve — APPROVED** [review]
Great work! :)

**Pull request merged**